.github/workflows/checks.yml

```diff
@@ -8,49 +8,108 @@ on:
   pull_request:
 
 jobs:
-  build:
+  setup:
+    name: Download deps, install poetry, populate caches
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ["3.8", "3.9"]
+        python-version: [ "3.8", "3.9" ]
     steps:
-    - uses: actions/checkout@v3
-      with:
-        submodules: recursive
-    - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v4
-      with:
-        python-version: ${{ matrix.python-version }}
-    - name: Install poetry
-      run: |
-        python -m pip install --upgrade pip
-        pip install poetry
-    - name: Check poetry.lock
-      run: |
-        poetry lock --check
-    - name: Install dependencies
-      run: |
-        poetry install
-    - name: Check formatting with black
-      run: |
-        poetry run poe format_check
-    - name: Analysing the code with pylint
-      run: |
-        poetry run poe lint
-    - name: Typechecking
-      run: |
-        poetry run poe typecheck
-    - name: Check if requirements.txt is up-to-date
-      run: |
-        poetry export -f requirements.txt --without-hashes --extras docs | cmp - requirements.txt
-    - name: Compile contracts
-      run: |
-        poetry run poe compile_contracts
-    - name: Unit & e2e test
-      run: |
-        poetry run poe test_ci
-    - name: Generate coverage in XML
-      run: |
-        poetry run coverage xml
-    - name: Upload coverage to Codecov
-      uses: codecov/codecov-action@v3
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install poetry
+        run: |
+          python -m pip install --upgrade pip
+          pip install poetry
+      - name: Check poetry.lock
+        run: |
+          poetry lock --check
+      - name: Install dependencies
+        run: |
+          poetry install
+      - name: Populate caches
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cache/pypoetry
+          key: poetry-${{ hashFiles('poetry.lock') }}
+
+  lint:
+    name: Check formatting, linting and types
+    runs-on: ubuntu-20.04
+    needs: setup
+    strategy:
+      matrix:
+        python-version: [ "3.8", "3.9" ]
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
```

Comment:
Why do we set up python in every job? Is setting it in `setup` not enough?

Reply:
Nope, it has to be set up for every job